[macOS] Spurious ASSERT fails under WebCore::DictionaryLookup::stringForPDFSelection with empty selection

aprotyas opened the pull request:
#### 16048339c64acda4ebd00a9caec15b9cc3b7ab89
<pre>
[macOS] Spurious ASSERT fails under WebCore::DictionaryLookup::stringForPDFSelection with empty selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=283847">https://bugs.webkit.org/show_bug.cgi?id=283847</a>
<a href="https://rdar.apple.com/128668294">rdar://128668294</a>

Reviewed by Wenson Hsieh.

On some documents, we frequently ASSERT during lookup hit tests under
DictionaryLookup::stringForPDFSelection when clicking on (essentially)
blank text. This is because selection extension code in that function
can also end up modifying a PDF selection to nil if the Reveal framework
says the extracted lookup text range under the selection is empty. This
is problematic because the plain text on the text range in question is
an empty string, rather than nil, so our isEqualToString: assertion
fails.

This specific failure mode is unexpected, so we deflake the assertion in
two ways:

1. When the PDF selection is nil, instead of an NSString equality check,
   we check if the plain text on the text range is empty. This is a
   string equality check where we ignore nullity on one side. We
   fallback to the full isEqualToString: check if the PDF selection is
   not nil.
2. Augment an early return in UnifiedPDFPlugin::textForImmediateActionHitTestAtPoint()
   with a check for an empty PDF selection. This way, we are not
   unnecessarily performing lookup on empty text ranges.

To add testing coverage, we move the WKWebViewForTestingImmediateActions
class into a separate namesake header, such that it can be used in the
Unified PDF test suite.

* Source/WebCore/editing/cocoa/DictionaryLookup.mm:
(WebCore::DictionaryLookup::stringForPDFSelection):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::textForImmediateActionHitTestAtPoint):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/metalSpecTOC.pdf: Added.
* Tools/TestWebKitAPI/Tests/mac/ImmediateActionTests.mm:
(swizzledImmediateActionLocationInView): Deleted.
(-[WKWebViewForTestingImmediateActions _immediateActionAnimationControllerForHitTestResult:withType:userData:]): Deleted.
(-[WKWebViewForTestingImmediateActions immediateActionGesture]): Deleted.
(-[WKWebViewForTestingImmediateActions simulateImmediateAction:]): Deleted.
* Tools/TestWebKitAPI/Tests/mac/WKWebViewForTestingImmediateActions.h: Added.
* Tools/TestWebKitAPI/Tests/mac/WKWebViewForTestingImmediateActions.mm: Added.
(swizzledImmediateActionLocationInView):
(-[WKWebViewForTestingImmediateActions _immediateActionAnimationControllerForHitTestResult:withType:userData:]):
(-[WKWebViewForTestingImmediateActions immediateActionGesture]):
(-[WKWebViewForTestingImmediateActions simulateImmediateAction:]):

Canonical link: <a href="https://commits.webkit.org/287236@main">https://commits.webkit.org/287236@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3d76bdba06a5cd8db59995fa0cbdd37676e17e0e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78886 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57929 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32267 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83541 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30141 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81019 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67074 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6210 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61773 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19703 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81953 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51816 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/71775 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42076 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/49166 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25977 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28485 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/70280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26386 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84915 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6246 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69998 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6408 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67809 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69249 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17246 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13296 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/12039 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6192 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6172 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9611 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7963 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Tools/TestWebKitAPI/Tests/mac/WKWebViewForTestingImmediateActions.mm:
```diff
@@ -0,0 +1,96 @@
+{
+    for (NSGestureRecognizer *gesture in [self gestureRecognizers]) {
+        if ([gesture isKindOfClass:NSImmediateActionGestureRecognizer.class])
+            return static_cast<NSImmediateActionGestureRecognizer *>(gesture);
```
Nit - we can probably patch this up to use `dynamic_objc_cast` while we're moving it.

```suggestion
        if (RetainPtr immediateActionGesture = dynamic_objc_cast<NSImmediateActionGestureRecognizer>(gesture))
            return immediateActionGesture.get();
```

rr-codes commented on Source/WebCore/editing/cocoa/DictionaryLookup.mm:
```diff
@@ -411,8 +411,14 @@ static void expandSelectionByCharacters(PDFSelection *selection, NSInteger numbe
-    return selection.string;
+    NSString *selectionString = selection.string;
+
+    auto extractedStringMatchesSelection = [](auto selection, auto extracted) -> bool {
```
this can probably be guarded by ASSERT_ENABLED so that it's not unnecessary performed

whsieh replied:
I don't think this logic will be run on release builds, since the only call to it would be compiled out altogether.

aprotyas replied:
Right

whsieh commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3366,8 +3366,8 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
-    if (!wordSelection)
-        return { { }, nil };
+    if (!wordSelection || [wordSelection isEmpty])
+        return { { }, wordSelection ?: nil };
```
Wouldn't `wordSelection ?: nil` just be equivalent to `wordSelection`?

```suggestion
        return { { }, wordSelection };
```

aprotyas replied:
😅 of course

aprotyas committed:
[macOS] Spurious ASSERT fails under WebCore::DictionaryLookup::stringForPDFSelection with empty selection
https://bugs.webkit.org/show_bug.cgi?id=283847
rdar://128668294

Reviewed by Wenson Hsieh.

On some documents, we frequently ASSERT during lookup hit tests under
DictionaryLookup::stringForPDFSelection when clicking on (essentially)
blank text. This is because selection extension code in that function
can also end up modifying a PDF selection to nil if the Reveal framework
says the extracted lookup text range under the selection is empty. This
is problematic because the plain text on the text range in question is
an empty string, rather than nil, so our isEqualToString: assertion
fails.

This specific failure mode is unexpected, so we deflake the assertion in
two ways:

1. When the PDF selection is nil, instead of an NSString equality check,
   we check if the plain text on the text range is empty. This is a
   string equality check where we ignore nullity on one side. We
   fallback to the full isEqualToString: check if the PDF selection is
   not nil.
2. Augment an early return in UnifiedPDFPlugin::textForImmediateActionHitTestAtPoint()
   with a check for an empty PDF selection. This way, we are not
   unnecessarily performing lookup on empty text ranges.

To add testing coverage, we move the WKWebViewForTestingImmediateActions
class into a separate namesake header, such that it can be used in the
Unified PDF test suite.

* Source/WebCore/editing/cocoa/DictionaryLookup.mm:
(WebCore::DictionaryLookup::stringForPDFSelection):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::textForImmediateActionHitTestAtPoint):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/metalSpecTOC.pdf: Added.
* Tools/TestWebKitAPI/Tests/mac/ImmediateActionTests.mm:
(swizzledImmediateActionLocationInView): Deleted.
(-[WKWebViewForTestingImmediateActions _immediateActionAnimationControllerForHitTestResult:withType:userData:]): Deleted.
(-[WKWebViewForTestingImmediateActions immediateActionGesture]): Deleted.
(-[WKWebViewForTestingImmediateActions simulateImmediateAction:]): Deleted.
* Tools/TestWebKitAPI/Tests/mac/WKWebViewForTestingImmediateActions.h: Added.
* Tools/TestWebKitAPI/Tests/mac/WKWebViewForTestingImmediateActions.mm: Added.
(swizzledImmediateActionLocationInView):
(-[WKWebViewForTestingImmediateActions _immediateActionAnimationControllerForHitTestResult:withType:userData:]):
(-[WKWebViewForTestingImmediateActions immediateActionGesture]):
(-[WKWebViewForTestingImmediateActions simulateImmediateAction:]):

Canonical link: https://commits.webkit.org/287236@main
